tests/e2e-kubernetes/testsuites/credentials.go

```diff
@@ -223,287 +239,311 @@ func (t *s3CSICredentialsTestSuite) DefineTests(driver storageframework.TestDriv
+					deferCleanup(restoreServiceAccountRole)
 
-			It("should use service account's read-only role", func(ctx context.Context) {
-				updateServiceAccountRole(ctx, iamPolicyS3ReadOnlyAccess)
-				pod := createPodWithVolume(ctx)
-				expectReadOnly(pod)
-			})
+					// Trigger recreation of our pods to use the new IAM role
+					killCSIDriverPods(ctx, f)
+				}
 
-			It("should use service account's full access role", func(ctx context.Context) {
-				updateServiceAccountRole(ctx, iamPolicyS3FullAccess)
-				pod := createPodAllowsDelete(ctx)
-				expectFullAccess(pod)
+				It("should use service account's read-only role", func(ctx context.Context) {
+					updateServiceAccountRole(ctx, iamPolicyS3ReadOnlyAccess)
+					pod := createPodWithVolume(ctx)
+					expectReadOnly(pod)
+				})
+
+				It("should use service account's full access role", func(ctx context.Context) {
+					updateServiceAccountRole(ctx, iamPolicyS3FullAccess)
+					pod := createPodAllowsDelete(ctx)
+					expectFullAccess(pod)
+				})
+
+				It("should fail to mount if service account's role does not allow s3::ListObjectsV2", func(ctx context.Context) {
+					updateServiceAccountRole(ctx, iamPolicyS3NoAccess)
+					expectFailToMount(ctx, "")
+				})
 			})
 
-			It("should fail to mount if service account's role does not allow s3::ListObjectsV2", func(ctx context.Context) {
-				updateServiceAccountRole(ctx, iamPolicyS3NoAccess)
-				expectFailToMount(ctx)
+			Context("Credentials via Kubernetes Secrets", func() {
+				updateCredentials := func(ctx context.Context, policyARN string) {
+					By("Updating Kubernetes Secret with temporary credentials")
+
+					role, ok := policyRoleMapping[policyARN]
+					if !ok {
+						framework.Failf("Missing role mapping for policy %s", policyARN)
+					}
+					assumeRoleOutput := assumeRole(ctx, f, *role.Arn)
+
+					_, deleteSecret := createCredentialSecret(ctx, f, assumeRoleOutput.Credentials)
+					deferCleanup(deleteSecret)
+
+					// Trigger recreation of our pods to use the new credentials
+					killCSIDriverPods(ctx, f)
+				}
+
+				It("should use read-only access aws credentials", func(ctx context.Context) {
+					updateCredentials(ctx, iamPolicyS3ReadOnlyAccess)
+					pod := createPodWithVolume(ctx)
+					expectReadOnly(pod)
+				})
+
+				It("should use full access aws credentials", func(ctx context.Context) {
+					updateCredentials(ctx, iamPolicyS3FullAccess)
+					pod := createPodAllowsDelete(ctx)
+					expectFullAccess(pod)
+				})
+
+				It("should fail to mount if aws credentials does not allow s3::ListObjectsV2", func(ctx context.Context) {
+					updateCredentials(ctx, iamPolicyS3NoAccess)
+					expectFailToMount(ctx, "")
+				})
 			})
 		})
 
-		Context("Credentials via Kubernetes Secrets", func() {
-			updateCredentials := func(ctx context.Context, policyARN string) {
-				By("Updating Kubernetes Secret with temporary credentials")
+		Describe("Pod level", func() {
```

Comment:
Could you do a commit without this added and one with? Doing a full review with changes from all commits makes this really hard to read as the indentation for the other stuff changed here

Reply:
Can we also have tests for:
- SA attached to pod doesn't have any annotation, and is marked to use pod credentials -> fails to mount
- SA attached to pod has invalid annotation (invalid ARN/non-existant ARN/ARN without a trust policy set), and is marked to use pod credentials -> fails to mount

Reply:
https://github.com/awslabs/mountpoint-s3-csi-driver/pull/235/commits/b579efbb1656113be1745ee1cb09b7cd2025a0c3 this commit only includes pod-level tests